Support custom package server

mint-thompson opened the pull request:
Part of task [CIMPL-1000](https://standardhealthrecord.atlassian.net/browse/CIMPL-1000).

If the FPL_CUSTOM_REGISTRY environment variable is set, use it when downloading dependencies. This is useful if you are on a corporate network where you have to use a different package registry.

mint-thompson committed:
Support custom package server

If the FPL_CUSTOM_REGISTRY environment variable is set, use it when
downloading dependencies.

cmoesel commented on src/load.ts:
```diff
@@ -211,7 +212,11 @@ export async function mergeDependency(
   } else if (!loadedPackage) {
-    packageUrl = `https://packages.fhir.org/${packageName}/${version}`;
+    if (process.env.FPL_CUSTOM_REGISTRY) {
+      packageUrl = `${process.env.FPL_CUSTOM_REGISTRY}/${packageName}/${version}`;
```
If `FPL_CUSTOM_REGISTRY` ends with `/`, then there will be a double-`/`. I think we should support a registry URL with or without a trailing slash, so if there is one there, we should probably trim it off.

cmoesel reviewed (changes requested):
This is a lot cleaner than I expected it would be! Just one spot! Well... we can't have that! ;-)

I think we want to make it somewhat obvious when a custom registry is being used -- so I'd like to log an INFO near the beginning that says something like:
```
info  Using custom registry specified by FPL_CUSTOM_REGISTRY environment variable: http://foo.org/fhir
```

I'd also suggest that when it's not found, instead of saying
> could not be loaded locally or from the FHIR package registry

say
> could not be loaded locally or from the custom FHIR package registry http://foo.org/fhir

Last -- not a super big deal, but why not just `FPL_REGISTRY`? I don't think `CUSTOM` is needed...

mint-thompson committed:
Rename environment variable to FPL_REGISTRY

The custom registry is checked in a helper function. The helper function
will try to log information about a custom registry the first time it is
called.

If the custom registry has a trailing slash, remove it before using it
to make a package URL.

mint-thompson commented:
@cmoesel Thank you for the review comments! I've made updates based on your suggestions. The only potential unusual situation is that it might be possible for the info message in `getCustomRegistry` to be logged more than once due to race conditions, but since the only effect of that is an extra info message, I don't think it's necessary to add a mutex or anything like that.

jafeltra commented on test/load.test.ts:
```diff
@@ -584,6 +586,7 @@ describe('#mergeDependency()', () => {
     removeSpy = jest.spyOn(fs, 'removeSync').mockImplementation(() => {});
     moveSpy = jest.spyOn(fs, 'moveSync').mockImplementation(() => {});
     cachePath = path.join(__dirname, 'testhelpers', 'fixtures');
+    delete process.env.FPL_REGISTRY;
```
Is this to ensure any environment variable set locally are not used?

mint-thompson replied:
Yes. If the tests are running on a system where the environment variable is set, we don't want that to affect the tests.

jafeltra commented on src/utils/customRegistry.ts:
```diff
@@ -0,0 +1,17 @@
+export function getCustomRegistry(log: LogFunction = () => {}) {
+  if (process.env.FPL_REGISTRY) {
+    if (!hasLoggedCustomRegistry) {
+      hasLoggedCustomRegistry = true;
```
I wonder if instead of using the flag and possibly have in the race condition, could the info message just be logged after [line 216](https://github.com/standardhealth/fhir-package-loader/pull/10/files#diff-51cd61f2399d311debdd2c813260b4a799b0cd6155b8eaa0f1498aebdc5ca281R216) before the `packageUrl` with the custom registry is set?

cmoesel replied:
Node.js is single threaded and there aren't even any async calls between checking and setting the variable. I'm not sure we need to worry about this.

mint-thompson replied:
If the message were logged there, it would get logged each time `mergeDependency` is called. The purpose of this implementation is that it only logs the message once, no matter how many times `mergeDependency` is called.

jafeltra replied:
Ohh right. I lost track of where that function was being called from. Thanks!

jafeltra reviewed (commented):
This looks really good! I just left two really small questions for my own understanding. I like how simple this turned out!

cmoesel reviewed (approved):
I like it! Thanks for the updates. I tested everything out manually and it all looks good to me!
